packages/core/mesh/network-manager/src/transport/libdatachannel-transport.ts

```diff
@@ -0,0 +1,241 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { Duplex } from 'stream';
+
+import { Event, Trigger } from '@dxos/async';
+import { ErrorStream } from '@dxos/debug';
+import { log } from '@dxos/log';
+import { Signal } from '@dxos/protocols/proto/dxos/mesh/swarm';
+
+import { type PeerConnection } from './datachannel/rtc-peer-connection';
+import { Transport } from './transport';
+
+export type LibDataChannelTransportParams = {
+  initiator: boolean;
+  stream: NodeJS.ReadWriteStream;
+  webrtcConfig?: RTCConfiguration;
+  sendSignal: (signal: Signal) => Promise<void>;
+};
+
+const MAX_BUFFERED_AMOUNT = 64 * 1024;
+
+// https://viblast.com/blog/2015/2/5/webrtc-data-channel-message-size/
+const MAX_MESSAGE_SIZE = 64 * 1024;
+
+/**
+ * Transport
+ */
+export class LibDataChannelTransport implements Transport {
+  private _closed = false;
+  readonly closed = new Event();
+  private _connected = false;
+  readonly connected = new Event();
+  readonly errors = new ErrorStream();
+  private readonly _peer: Promise<PeerConnection>;
+  private _channel!: RTCDataChannel;
+  private _stream!: Duplex;
+
+  private readonly _readyForCandidates = new Trigger();
+  private _writeCallback: (() => void) | null = null;
+
+  constructor(private readonly params: LibDataChannelTransportParams) {
+    this._peer = (async () => {
+      const { RTCPeerConnection: PeerConnection } = await import('./datachannel');
+      if (this._closed) {
+        this.errors.raise(Error('connection already closed'));
+      }
+      const peer = new PeerConnection(params.webrtcConfig);
+
+      peer.onicecandidateerror = (event) => {
+        log.error('peer.onicecandidateerror', { event });
+      };
+
+      peer.onconnectionstatechange = (event) => {
+        log.debug('peer.onconnectionstatechange', {
```

Comment:
its a code style issue but `log(....)` and `log.debug(...)` do the same thing - we usually just use `log(..)` in the rest of the codebase

Reply:
understanding your comment, but that's a bit astonishing to me, i'd expect `log(` to log at whatever the current loglevel is set at, or at some level less chatty than `debug`.

i'm fine sticking with what the precedent and convention is, but we might consider documenting it and the rationale

Reply:
sure, happy to discuss this at some point